config/deploy/staging.rb

```diff
@@ -1,9 +1,38 @@
-# using an ssh-config entry to set the user, key, and address of the machine
+#Using the aws-sdk to locate servers via the Role tag and deploy to them.
+require 'aws-sdk-ec2'
+require 'yaml'
+require 'aws-sdk-core'
 set :stage, :staging
 set :rails_env, 'production'
-server 'kithe.sciencehistory.org', user: 'digcol', roles: [:web, :app, :db, :jobs, :solr, :cron]
-#server '34.227.26.131', user: 'hydep', roles: [:solr]
-#server '54.204.60.1', user: 'hydep', roles: [:jobs, :cron] # dzi bg job box
+secret_location = './aws_secrets.yml'
+server_roles = {"kithe"=>[':web', ':app', ':db', ':jobs', ':solr', ':cron']}
+aws_region = "us-east-1"
+service_level = "stage"
+#Everything below here should be able to be turned into a method, the variables above may change based on server setup.
+credential_file = File.file?(secret_location)
+#Checking for the needed credential file, which should overwrite any other ENV or file settings.
+if credential_file && true
+#Edit ec2 for region changes, right now we only use one region. Also needs to come after the credential steps otherwise [default] aws credentials in the .aws directory may be used if present leading to erratic behavior.
+  creds= YAML.load_file(secret_location)
+  Aws.config[:credentials] = Aws::Credentials.new(creds['AccessKeyId'],creds['SecretAccessKey'])
+  ec2 = Aws::EC2::Resource.new(region:"#{aws_region}")
+#Server role keys should be the Role tag (assigned by ansible) that you want to deploy to. The array value is the list of capistrano roles that the server needs.
+  server_roles.each do |key, value|
+#Service level is set manually here, maybe make it a variable further up to be easy to spot when making new stages? 
+#The instance-state-code of 16 is a value from Amazon's docs for a running server. See: https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstances.html
+    ec2.instances({filters: [{name:'instance-state-code', values:["16"]},{name: 'tag:Role', values: ["#{key}"]},{name: 'tag:Service_level', values: ["#{service_level}"]}]}).each do |aws_server|
+#Search across the tags and find the one labeled capistrano_roles, tags are hashes with 2 values, key for tag name and value for tag value.
```

Comment:
@sanfordd Did we talk about putting in another tag that limited to the "kithe" application, like `tag:application` maybe?  In case we start using this system for other applications, we only want to pull back EC2 hosts for _this_ one.  Or is that there and I'm missing it? (although I'd rather not call the value `kithe`, that is not the name of this application. We could call call it `digicol2` maybe?)

Reply:
Oh wait, i see the `key` variable is that. Can you change the name to `tag:Application` instead of `tag:Role`, and use a value that describes our application rather than `kithe` which is just one dependency of it (which could hypothetically change)?

I also notice you are using uppercase-first-letter for some tags `tag:Role`, `tag:Service`, but lowercase for "capistrano_roles" below. Make em consistent?

Reply:
I'd prefer not to use numbers in the application name since we're also using them for the specific server identification digcol2-1, 2-2, 2-3 seems likely to introduce some confusion versus name-1, name-2. digcol (or digicoll as we current have the repo named) would be fine.

Reply:
Makes sense, I was just worried about confusing it with our current chf_sufia app which is also "digital collections", this new one is "version 2". We could come up with a snazzy internal codename for it, like Alberta calls theirs "jupiter" but that doens't seem likely to reduce confusion. I just don't like "kithe", becuase it's an implementation detail which could change. If you think `digicoll` is fine, fine with me. 

Also rename "Role" to "Application" makes sense for how we're now going to use it?

Reply:
I'll add an application section, we do want to keep Role (or a tag akin to it) so we can distinguish between a solr server, web server, jobs server, or other specific purpose server in group of servers running the new application.

Reply:
> so we can distinguish between a solr server, web server, jobs server, or other specific purpose server in group of servers running the new application.

As far as capistrano is concerned, that's what `capistrano_role` defines, no?  If you need the role for other purposes not related to capistrano, that's cool (although I don't understand what a role "kithe" means, but I don't know if I need to) -- but I don't think capistrano needs to know about it. Capistrano needs to know about all the servers in the "application" of concern, at the "service_level" of concern -- and then it needs to know what capistrano_roles they have, because that's how it knows how/what to deploy to each one. I think that's all capistrano needs to know about, not this additional "Tag:Role".  Am I missing anything?

Reply:
The Application tag has been added and tested, and all tags are now following the convention of a capitalized first letter. Switching from prod/stage to production/staging is still in progress. There's a number of things that use those values.

Reply:
@jrochkind If want to get this out sooner, the service_level tag should be easy enough to adjust in the future by swapping out a variable. Otherwise if there's no time pressure I can keep adjusting service_level naming schemes.

Reply:
I'll work on adjusting it for staging to test with, due to work on the new server setup this will be delayed until next week.

Reply:
The renaming worked and we're using the full production/staging terminology.